{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.datasets import load_iris\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from typing import List, Optional, Callable, Tuple\n",
    "from numpy import random\n",
    "import itertools\n",
    "import math\n",
    "import statistics\n",
    "from tabulate import tabulate\n",
    "import time\n",
    "\n",
    "from sklearn.metrics import r2_score\n",
    "from sklearn.metrics import mean_squared_error as mse_score\n",
    "from sklearn.metrics import mean_absolute_error as mae_score\n",
    "from sklearn.metrics import mean_absolute_percentage_error as mape_score"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font size = 6> Genetic algo with Adaptive Mutation</font>\n",
    "\n",
    "\n",
    "Fitness based adaptive Mutation<br>\n",
    "https://neptune.ai/blog/adaptive-mutation-in-genetic-algorithm-with-python-examples<br>\n",
    "https://www.atlantis-press.com/article/7393.pdf\n",
    "\n",
    "Rank based adaptive mutation <br>\n",
    "https://arxiv.org/pdf/2104.08842\n",
    "\n",
    "A good crossover operator for real encoding was proposed by K. Deb in 1995. Here is the paper: http://www.complex-systems.com/pdf/09-2-2.pdf\n",
    "\n",
    "Parameters \n",
    "1. Param Grid: Given\n",
    "2. Genome: One combination of Param Grid\n",
    "3. Population: All combinations of Param Grid considered\n",
    "\n",
    "Functions\n",
    "1. Genetic Representation of Genome, Population (solutions)\n",
    "2. Generating new solutions\n",
    "3. Fitness Value functions: MSE, MAE, etc\n",
    "4. Elitism/Selection Function: Choose which Genome to undergo CrossOver, can use python's choice function, weight = Fitness Value\n",
    "5. Cross over Function\n",
    "6. Mutation Algo: Usually mutation rate is set to 1/L, where L is the length of the bitstring.\n",
    "7. Evolution Algo\n",
    "\n",
    "Classes\n",
    "1. Genome\n",
    "2. Population\n",
    "3. GenomSearch\n",
    "    - GenomSearchCV\n",
    "\n",
    "\n",
    "Limitations\n",
    "1. Must be a range\n",
    "2. cannot pipeline\n",
    "\n",
    "Future:\n",
    "1. Allow for more flexible formatting of param_grid\n",
    "2. Take care of edge cases\n",
    "3. Convert to binary (maybe faster?)\n",
    "4. In population.mutation() -> Maybe can update f_max and f_avg as we are mutating, else can try parallel processing\n",
    "\n",
    "Fitness Function:\n",
    "Fine-tune the fitness function to better represent the problem's objectives and constraints. A well-designed fitness function guides the algorithm toward the desired solutions.\n",
    "\n",
    "Constraint Handling:\n",
    "Implement techniques for handling constraints effectively. Penalization or repair methods can ensure that generated solutions comply with the problem constraints.\n",
    "\n",
    "Diversity Preservation:\n",
    "Encourage diversity within the population to prevent premature convergence to local optima. Diversity-preserving techniques, such as crowding or fitness sharing, can help in maintaining a diverse set of solutions.\n",
    "\n",
    "Hybrid Approaches:\n",
    "Consider combining the genetic algorithm with other optimization techniques, such as local search or simulated annealing, to explore the search space more effectively.\n",
    "\n",
    "Parameter Tuning:\n",
    "Perform parameter tuning using techniques like grid search or genetic algorithms themselves to optimize the genetic algorithm's parameters for your specific problem.\n",
    "\n",
    "Parallelization:\n",
    "If your problem domain allows it, consider parallelizing the genetic algorithm to run multiple populations concurrently. This can significantly speed up the search process.\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\nInput example:\\n\\nparams_grid = {\\n    \\'n\\': {range: [low, high], dtype: int}   ## (Low, High) represents the range, dtype = int, float etc\\n    \\'objective\\': (range: [cat1, cat2, cat3,...], dtype: category }\\n    \\'gamma\\': {range: [low, high], dtype: float}\\n}\\n\\navailable dtypes:\\ncategory, int, float\\n\\npopulation = [Genome1, Genome2, Genome3, ...]\\n\\n\\nneed to make one for pipelining so like \"PCA__n componenes\"\\n'"
      ]
     },
     "execution_count": 146,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Input example:\n",
    "\n",
    "params_grid = {\n",
    "    'n': {range: [low, high], dtype: int}   ## (Low, High) represents the range, dtype = int, float etc\n",
    "    'objective': (range: [cat1, cat2, cat3,...], dtype: category }\n",
    "    'gamma': {range: [low, high], dtype: float}\n",
    "}\n",
    "\n",
    "available dtypes:\n",
    "category, int, float\n",
    "\n",
    "population = [Genome1, Genome2, Genome3, ...]\n",
    "\n",
    "\n",
    "need to make one for pipelining so like \"PCA__n componenes\"\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Helpful functions:\n",
    "1. Timsort, insertion, merge\n",
    "2. Merge dictionaries\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "MIN_MERGE = 32\n",
    "\n",
    "\n",
    "def calcMinRun(n):\n",
    "    \"\"\"Returns the minimum length of a\n",
    "    run from 23 - 64 so that\n",
    "    the len(array)/minrun is less than or\n",
    "    equal to a power of 2.\n",
    "\n",
    "    e.g. 1=>1, ..., 63=>63, 64=>32, 65=>33,\n",
    "    ..., 127=>64, 128=>32, ...\n",
    "    \"\"\"\n",
    "    r = 0\n",
    "    while n >= MIN_MERGE:\n",
    "        r |= n & 1\n",
    "        n >>= 1\n",
    "    return n + r\n",
    "\n",
    "\n",
    "# This function sorts array from left index to\n",
    "# to right index which is of size atmost RUN\n",
    "def insertionSort(pop, left, right, dir):\n",
    "    for i in range(left + 1, right + 1):\n",
    "        j = i\n",
    "\n",
    "        if dir == 1:       ## Ascendimg\n",
    "            while j > left and pop[j] < pop[j - 1]:\n",
    "                pop[j], pop[j - 1] = pop[j - 1], pop[j]\n",
    "                j -= 1\n",
    "        else:               ## Descending\n",
    "            while j > left and pop[j] > pop[j - 1]:\n",
    "                pop[j], pop[j - 1] = pop[j - 1], pop[j]\n",
    "                j -= 1\n",
    "\n",
    "\n",
    "# Merge function merges the sorted runs\n",
    "def merge(pop, l, m, r, dir = 1):\n",
    "\n",
    "    # original array is broken in two parts\n",
    "    # left and right array\n",
    "    len1, len2 = m - l + 1, r - m\n",
    "    left, right = [], []\n",
    "    for i in range(0, len1):\n",
    "        left.append(pop[l + i])\n",
    "    for i in range(0, len2):\n",
    "        right.append(pop[m + 1 + i])\n",
    "\n",
    "    i, j, k = 0, 0, l\n",
    "\n",
    "    # after comparing, we merge those two array\n",
    "    # in larger sub array\n",
    "    if dir == 1:   ## Ascending\n",
    "        while i < len1 and j < len2:\n",
    "            if left[i] <= right[j]:\n",
    "                pop[k] = left[i]\n",
    "                i += 1\n",
    "\n",
    "            else:\n",
    "                pop[k] = right[j]\n",
    "                j += 1\n",
    "\n",
    "            k += 1\n",
    "    else:           ## Descending\n",
    "        while i < len1 and j < len2:\n",
    "            if left[i] >= right[j]:\n",
    "                pop[k] = left[i]\n",
    "                i += 1\n",
    "\n",
    "            else:\n",
    "                pop[k] = right[j]\n",
    "                j += 1\n",
    "\n",
    "            k += 1\n",
    "\n",
    "    # Copy remaining elements of left, if any\n",
    "    while i < len1:\n",
    "        pop[k] = left[i]\n",
    "        k += 1\n",
    "        i += 1\n",
    "\n",
    "    # Copy remaining element of right, if any\n",
    "    while j < len2:\n",
    "        pop[k] = right[j]\n",
    "        k += 1\n",
    "        j += 1\n",
    "\n",
    "# Iterative Timsort function to sort the\n",
    "# array[0...n-1] (similar to merge sort)\n",
    "def tim_sort(pop, dir = 1):\n",
    "    \"\"\"\n",
    "    Purpose:\n",
    "    Sorts population based on fitness values\n",
    "\n",
    "    Input: \n",
    "    pop: Takes in a population\n",
    "    dir: '1' - Ascending, '0' - 'Descending'\n",
    "    \"\"\"\n",
    "    n = len(pop)\n",
    "    minRun = calcMinRun(n)\n",
    "\n",
    "    # Sort individual subarrays of size RUN\n",
    "    for start in range(0, n, minRun):\n",
    "        end = min(start + minRun - 1, n - 1)\n",
    "        insertionSort(pop, start, end, dir)\n",
    "\n",
    "    # Start merging from size RUN (or 32). It will merge\n",
    "    # to form size 64, then 128, 256 and so on ....\n",
    "    size = minRun\n",
    "    while size < n:\n",
    "\n",
    "        # Pick starting point of left sub array. We\n",
    "        # are going to merge arr[left..left+size-1]\n",
    "        # and arr[left+size, left+2*size-1]\n",
    "        # After every merge, we increase left by 2*size\n",
    "        for left in range(0, n, 2 * size):\n",
    "\n",
    "            # Find ending point of left sub array\n",
    "            # mid+1 is starting point of right sub array\n",
    "            mid = min(n - 1, left + size - 1)\n",
    "            right = min((left + 2 * size - 1), (n - 1))\n",
    "\n",
    "            # Merge sub array arr[left.....mid] &\n",
    "            # arr[mid+1....right]\n",
    "            if mid < right:\n",
    "                merge(pop, left, mid, right)\n",
    "\n",
    "        size = 2 * size\n",
    "\n",
    "    return pop\n",
    "\n",
    "def merge_dicts(dict1, dict2):\n",
    "    # Check for overlapping keys\n",
    "    intersection = set(dict1.keys()) & set(dict2.keys())\n",
    "    if intersection:\n",
    "        raise ValueError(f\"Overlapping keys found: {intersection}\")\n",
    "\n",
    "    # Merge the dictionaries\n",
    "    merged_dict = dict1.copy()\n",
    "    merged_dict.update(dict2)\n",
    "    \n",
    "    return merged_dict\n",
    "\n",
    "def check_error(param_grid):\n",
    "    \"\"\"\n",
    "    Purpose:\n",
    "    Just to check input\n",
    "    \"\"\"\n",
    "    avail_dtypes = ['int', 'float','category']\n",
    "\n",
    "    pass\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Genome():\n",
    "    def __init__(self, X_train, X_test, y_train, y_test, estimator, metric, params):\n",
    "        self.X_train = X_train\n",
    "        self.X_test = X_test\n",
    "        self.y_train = y_train\n",
    "        self.y_test = y_test\n",
    "        self._estimator = estimator\n",
    "        self.metric = metric\n",
    "        self._params = params\n",
    "        self._fitness = self.calc_fitness(params)\n",
    "        \n",
    "    \n",
    "    @property\n",
    "    def estimator(self):\n",
    "        return self._estimator\n",
    "\n",
    "    @property\n",
    "    def params(self):\n",
    "        return self._params\n",
    "\n",
    "    @property\n",
    "    def fitness(self):\n",
    "        return self._fitness\n",
    "\n",
    "    @params.setter\n",
    "    def params(self, new_params):\n",
    "        self._fitness = self.calc_fitness(new_params)\n",
    "        self._params = new_params\n",
    "\n",
    "\n",
    "    ## Comparing operators\n",
    "    def __lt__(self, other):\n",
    "        # Define custom behavior for \"<\"\n",
    "        return self.fitness < other.fitness\n",
    "\n",
    "    def __le__(self, other):\n",
    "        # Define custom behavior for \"<=\"\n",
    "        return self.fitness <= other.fitness\n",
    "\n",
    "    def __ge__(self, other):\n",
    "        # Define custom behavior for \">=\"\n",
    "        return self.fitness >= other.fitness\n",
    "\n",
    "    def __eq__(self, other):\n",
    "        # Define custom behavior for \"==\"\n",
    "        return self.fitness == other.fitness\n",
    "\n",
    "    def __ne__(self, other):\n",
    "        # Define custom behavior for \"!=\"\n",
    "        return self.fitness != other.fitness\n",
    "    \n",
    "    def __len__(self):\n",
    "        return len(self.params.keys())\n",
    "    \n",
    "    def __repr__(self):\n",
    "        string = f'Genome = Params : {self.params}, Fitness : {self.fitness}'\n",
    "        return string\n",
    "    \n",
    "    def extract(self, start, stop = None, step = 1):\n",
    "        \"\"\"\n",
    "        Just for ease of slicing\n",
    "        \n",
    "        Parameters\n",
    "        -----------------\n",
    "        key: Can be slice object\n",
    "        \"\"\"\n",
    "        if not stop: ## If stop == None, means only want extract one\n",
    "            return dict([list(self.params.items())[start]])\n",
    "            \n",
    "        elif stop > start and stop <= len(self):\n",
    "            target = dict([list(self.params.items())[i] for i in range(start, stop, step)])\n",
    "            return target\n",
    "        else:\n",
    "            raise IndexError(f\"Index out of range.\")\n",
    "\n",
    "    \n",
    "    def calc_fitness(self, params):\n",
    "        \"\"\"\n",
    "        Parameters\n",
    "        -------------------\n",
    "        Takes in 1 genome/set of params\n",
    "        \"\"\"\n",
    "\n",
    "        model = self.estimator\n",
    "\n",
    "        model.set_params(**params)\n",
    "\n",
    "        model.fit(self.X_train, self.y_train)\n",
    "\n",
    "        y_pred = model.predict(self.X_test)\n",
    "\n",
    "        results = self.evaluate(self.y_test, y_pred, self.metric)\n",
    "\n",
    "        return results\n",
    "\n",
    "    \n",
    "    def mutate(self, param_grid, rate):\n",
    "        \"\"\"\"\n",
    "        Mutate one gene\n",
    "\n",
    "        Parameters\n",
    "        --------------------\n",
    "        choices: Param_grid for choosing\n",
    "        rate: Mutation rate/ Probability of one parameter undergoing mutation\n",
    "\n",
    "        Returns\n",
    "        -------------------\n",
    "        Returns new gene\n",
    "        \"\"\"\n",
    "        print(f'original: {self}')\n",
    "\n",
    "        length = len(self)\n",
    "        choices = list(range(0, length))\n",
    "        params = self.params.copy()\n",
    "        keys = list(params.keys())\n",
    "\n",
    "        ################################\n",
    "        # Edge case 1: Only one choice, unable to mutate, so just remove from choices\n",
    "        for index, (key, items) in enumerate(param_grid.items()):\n",
    "            if len(items['range']) < 2:\n",
    "                   choices.remove(index)\n",
    "        ##################################\n",
    "\n",
    "        for i in choices:\n",
    "            prob = random.random()\n",
    "            if prob < rate:   ## If 0 < prob and prob < rate, means falls under the possibility of mutating\n",
    "                key = keys[i]\n",
    "                target_val = params[key]\n",
    "                items = param_grid[key]\n",
    "                new_target_val = target_val\n",
    "\n",
    "                while new_target_val == target_val:     ## Iterate until not the same\n",
    "                    if items['dtype'] == 'category':\n",
    "                        new_target_val = random.choice(items['range'])\n",
    "                    elif items['dtype'] == 'int':\n",
    "                        low, high = items['range']\n",
    "                        new_target_val = random.randint(low, high)\n",
    "                    elif items['dtype'] == 'float':\n",
    "                        low, high = items['range']\n",
    "                        new_target_val = round(random.uniform(low, high), 5)\n",
    "\n",
    "                params[key] = new_target_val\n",
    "\n",
    "                print(f\"After mutation: '{key}': '{new_target_val}'\")\n",
    "            else:\n",
    "                print(f'No mutation.')\n",
    "                continue\n",
    "\n",
    "        new_gene = Genome(self.X_train, self.X_test, self.y_train, self.y_test, self.estimator, self.metric, params)\n",
    "\n",
    "        return new_gene\n",
    "    \n",
    "    def evaluate(self, actual, predicted, metric):\n",
    "        '''Returns desired metrics'''\n",
    "\n",
    "        metrics = {\n",
    "                    \"r2\":r2_score(actual, predicted), \n",
    "                    \"rmse\":mse_score(actual, predicted)**0.5,\n",
    "                    \"mse\":mse_score(actual, predicted), \n",
    "                    \"mae\":mae_score(actual, predicted),\n",
    "                    }\n",
    "        return metrics[metric]\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Population():\n",
    "    def __init__(self, X_train, X_test, y_train, y_test, estimator, param_grid, pop_size, metric, population = None):\n",
    "        self._X_train = X_train\n",
    "        self._X_test = X_test\n",
    "        self._y_test = y_test\n",
    "        self._y_train = y_train\n",
    "        self._estimator = estimator\n",
    "        self._param_grid = param_grid\n",
    "        self._pop_size = pop_size\n",
    "        self._metric = metric\n",
    "        if not population:  ## If no population\n",
    "            self._population = self.generate_pop()\n",
    "        else:\n",
    "            self._population = population\n",
    "\n",
    "    @property\n",
    "    def X_train(self):\n",
    "        return self._X_train\n",
    "    \n",
    "    @property\n",
    "    def X_test(self):\n",
    "        return self._X_test\n",
    "    \n",
    "    @property\n",
    "    def y_train(self):\n",
    "        return self._y_train\n",
    "    \n",
    "    @property\n",
    "    def y_test(self):\n",
    "        return self._y_test\n",
    "\n",
    "    @property\n",
    "    def estimator(self):\n",
    "        return self._estimator\n",
    "    \n",
    "    @property\n",
    "    def population(self):\n",
    "        return tim_sort(self._population)\n",
    "    \n",
    "    @property\n",
    "    def param_grid(self):\n",
    "        return self._param_grid\n",
    "    \n",
    "    @property\n",
    "    def pop_size(self):\n",
    "        return self._pop_size\n",
    "    \n",
    "    @property\n",
    "    def metric(self):\n",
    "        return self._metric\n",
    "\n",
    "    @population.setter\n",
    "    def population(self, new_pop):\n",
    "        self._population = new_pop\n",
    "\n",
    "    ## Used this only for sorting so it's easier\n",
    "    def __getitem__(self, key):\n",
    "        if isinstance(key, int):\n",
    "            return self.population[key]\n",
    "        \n",
    "        elif isinstance(key, slice):\n",
    "            start, stop, step = key.indices(self.n)\n",
    "            target = [self.population[i] for i in range(start, stop, step)]\n",
    "            return target\n",
    "        else:\n",
    "            raise TypeError(f\"Invalid Argument type. Must be int or slice.\")\n",
    "        \n",
    "    def __repr__(self):\n",
    "        string = \"\"\n",
    "        for genome in self.population:\n",
    "            string += f'{genome}\\n'\n",
    "        return string\n",
    "        \n",
    "    def copy(self):\n",
    "        \"\"\"\n",
    "        Makes a copy of the current generation\n",
    "        \"\"\"\n",
    "        pop_copy = [genome for genome in self.population]\n",
    "        return Population(self.param_grid, self.pop_size, pop_copy)\n",
    "            \n",
    "    \n",
    "    def get_fit(self):\n",
    "        \"\"\"\n",
    "        Helps to get list of fitness values\n",
    "        \"\"\"\n",
    "        fits = []\n",
    "        for i in self:\n",
    "            fits += [i.fitness]\n",
    "        return fits\n",
    "\n",
    "    \n",
    "    def calc_avg_fit(self):\n",
    "        \"\"\"\n",
    "        For adaptive Mutation\n",
    "        \"\"\"\n",
    "        fits = self.get_fit()\n",
    "        return statistics.mean(fits)\n",
    "\n",
    "\n",
    "    def generate_pop(self):\n",
    "        random.seed(3)\n",
    "        \"\"\"\n",
    "        To initialize population\n",
    "\n",
    "        Parameters\n",
    "        --------------\n",
    "        n: Number of Genomes\n",
    "\n",
    "        Returns\n",
    "        ---------------\n",
    "        None\n",
    "\n",
    "        \"\"\"\n",
    "        pop = []\n",
    "        for i in range(self.pop_size):\n",
    "            params = {}\n",
    "            for key, items in self.param_grid.items():\n",
    "                if items['dtype'] == 'category':\n",
    "                    params[key] = random.choice(items['range'])\n",
    "                elif items['dtype'] == 'int':\n",
    "                    low, high = items['range']\n",
    "                    params[key] = random.randint(low, high)\n",
    "                else:\n",
    "                    low, high = items['range']\n",
    "                    params[key] = round(random.uniform(low, high), 5)\n",
    "            new_genome = Genome(self.X_train, self.X_test, self.y_train, self.y_test, self.estimator, self.metric, params)\n",
    "            pop += [new_genome]\n",
    "\n",
    "        return pop\n",
    "\n",
    "\n",
    "    def elitism(self, n):\n",
    "        \"\"\"\n",
    "        To choose which n Genomes to bring over to next gen\n",
    "\n",
    "        Parameters\n",
    "        ------------------\n",
    "        n: prop of Genomes, default 0.2 of population\n",
    "\n",
    "        Returns\n",
    "        -----------------\n",
    "        Top n Genomes\n",
    "        \"\"\"\n",
    "        if not 0 < n < 1:\n",
    "            raise ValueError(\"n must be between 0 and 1\")\n",
    "        n = round(n*self.pop_size)\n",
    "        top_n = self.population[0:n]\n",
    "\n",
    "        return top_n\n",
    "\n",
    "\n",
    "    def n_cross_over(self, genes, n, nco_rate = 0.5, verbose = True):\n",
    "        \"\"\"\n",
    "        To cross over 2 genomes\n",
    "        \n",
    "        Parameters\n",
    "        ------------------------\n",
    "        genes: (a,b) Cross over genomes in pop[a] and pop[b]\n",
    "        n: Number of points to cross over\n",
    "        verbose: Illustrate parents to genomes\n",
    "        nco_rate: Cross over rate/ Probability of genes undergoing crossover\n",
    "            -> If rate = 0: None of the genes can undergo mutation\n",
    "            -> If rate = 1: All of the genes can undergo mutation\n",
    "        \n",
    "        Returns\n",
    "        ----------------------\n",
    "        2 Genomes, Crossed over, not in place\n",
    "        \"\"\"\n",
    "        a,b = genes\n",
    "        a = int(a)\n",
    "        b = int(b)\n",
    "        gene1 = self.population[a]\n",
    "        gene2 = self.population[b]\n",
    "\n",
    "        len1 = len(gene1)\n",
    "        len2 = len(gene2)\n",
    "        if len1 != len2:\n",
    "            raise ValueError(\"Genomes must be of same length\")\n",
    "        \n",
    "        if len1 < 2:\n",
    "            raise Exception(\"Too short!\")\n",
    "        \n",
    "        if n >= len1 -1:\n",
    "            raise Exception(f\"Too much cross-over points. Maximum is {len1-1}.\")\n",
    "\n",
    "        prob = random.random()\n",
    "\n",
    "        if prob > nco_rate:  \n",
    "            print(\"p value = %.5f is more than %f. No cross over.\" %(prob, nco_rate))\n",
    "            return [gene1, gene2]\n",
    "\n",
    "        choices = list(range(1,len1-1))\n",
    "        idx = list(np.random.choice(choices, size = n, replace = False))\n",
    "        idxs = tim_sort(idx + [0, len1]) \n",
    "        sections = n+1\n",
    "        params3 = {}\n",
    "        params4 = {}\n",
    "\n",
    "        for i in range(sections):\n",
    "            ## Even\n",
    "            if i%2 == 0:\n",
    "                params33 = gene1.extract(idxs[i], idxs[i+1])\n",
    "                params44 = gene2.extract(idxs[i], idxs[i+1])\n",
    "                \n",
    "                    \n",
    "            ## Odd\n",
    "            else:\n",
    "                params33 = gene2.extract(idxs[i], idxs[i+1])\n",
    "                params44 = gene1.extract(idxs[i], idxs[i+1])\n",
    "            \n",
    "            params3 = merge_dicts(params3, params33)\n",
    "            params4 = merge_dicts(params4, params44)\n",
    "                \n",
    "        gene3 = Genome(self.X_train, self.X_test, self.y_train, self.y_test, self.estimator, self.metric, params3)\n",
    "        gene4 = Genome(self.X_train, self.X_test, self.y_train, self.y_test, self.estimator, self.metric, params4)\n",
    "\n",
    "        if verbose:\n",
    "            print(f'------------Parent Genomes------------\\n{gene1}\\n{gene2}\\n')\n",
    "            print(f'After {n}-point(s) crossover at index(es) : {idx}\\n')\n",
    "            print(f'------------Children Genomes-------------\\n{gene3}\\n{gene4}\\n')\n",
    "\n",
    "        return [gene3, gene4]\n",
    "\n",
    "\n",
    "    def selection(self, n, subset = None):\n",
    "        \"\"\"\"\n",
    "        Purpose:\n",
    "        Helps n_cross_over by choosing what 2 genomes to put in\n",
    "        \n",
    "        Input:\n",
    "        'n': Number of parent pairs\n",
    "        'subset': [low,high) Genes in population to select from eg. gene index 4:7 (exclusive)\n",
    "\n",
    "        Output:\n",
    "        List of tuples of indexes: [ (parent1,parent2) ,(parent3, parent4)]\n",
    "        \"\"\"\n",
    "        if not subset:\n",
    "            low, high = (0, self.pop_size)\n",
    "            size = self.pop_size\n",
    "        else:\n",
    "            low, high = subset\n",
    "            size = high -low \n",
    "\n",
    "        max = math.comb(size, 2)\n",
    "        if n > max:\n",
    "            raise ValueError(f'Too much combinations. Maximum is {max}')\n",
    "\n",
    "        pairs = []\n",
    "        choices = list(range(low, high))\n",
    "        fits = self.get_fit()[low:high]\n",
    "        p = [i/sum(fits) for i in fits]\n",
    "\n",
    "        for i in range(n):\n",
    "            parent1, parent2 = random.choice(choices, p = p, size = 2, replace = False)\n",
    "            pairs += [(parent1, parent2)]\n",
    "        \n",
    "        return pairs\n",
    "    \n",
    "    def fitness_mutation_rate(self, k, f_max, f_avg, f):\n",
    "        \"\"\"\n",
    "        Purpose:\n",
    "        Fitness-based adaptive mutation\n",
    "\n",
    "        Input:\n",
    "        k: (k1,k2) Tuple of 2 for constant, k1,k2 in (0,1)\n",
    "        f_max: Maximum fitness of population\n",
    "        f_avg: Average fitness of population\n",
    "        f: Current fitness value of Genome\n",
    "        \n",
    "        Output:\n",
    "        rate -> int\n",
    "        \"\"\"\n",
    "        k1, k2 = k\n",
    "        \n",
    "        if f >= f_avg:      ## High quality solution\n",
    "            rate = k1*( (f_max - f)/(f_max-f_avg) )\n",
    "        else:               ## Low quality solution\n",
    "            rate = k2\n",
    "        return rate\n",
    "\n",
    "    def rank_mutation_rate(self, p_max, r, n):\n",
    "        \"\"\"\"\n",
    "        Purpose:\n",
    "        Rank-based adaptive mutation\n",
    "        \n",
    "        Input:\n",
    "        p_max: Maximum mutation probability\n",
    "        r: Rank of chromosome\n",
    "        n: population size\n",
    "\n",
    "        Output:\n",
    "        rate -> int\n",
    "        \"\"\"\n",
    "        p = p_max*( 1- (r-1)/(n-1))\n",
    "        return p\n",
    "\n",
    "    def mutation(self, type, inplace = True, **params):\n",
    "        \"\"\"\n",
    "        Purpose:\n",
    "        Randomly select from population without replacement and mutate\n",
    "\n",
    "        Input:\n",
    "        type: 'fitneses', 'random', 'rank'\n",
    "            Default\n",
    "            - 'fitness' : k = (k1,k2) = (0.05,0.06)\n",
    "            - 'random': rate = 0.5\n",
    "            -'rank': p_max = 0.08\n",
    "        inplace: Mutate on spot\n",
    "        subset: [a, b) Genome at index a (inclusive) till b (exclusive)\n",
    "\n",
    "        Output:\n",
    "        Population with mutated genes, inplace\n",
    "\n",
    "        Warning:\n",
    "        Self.population has to be SORTED\n",
    "        \"\"\"\n",
    "        f_max = max(self.get_fit())\n",
    "        f_avg = self.calc_avg_fit()\n",
    "        new_pop = []\n",
    "\n",
    "        # if not subset:  \n",
    "        #     subset = self.population\n",
    "        # else: \n",
    "        #     a, b = subset\n",
    "        #     subset = self.population[a:b]\n",
    "\n",
    "        subset = self.population\n",
    "        ## Iterate through all the genome\n",
    "        for index, gene in enumerate(subset):\n",
    "            if type == 'fitness':\n",
    "                try:\n",
    "                    k = params['k']\n",
    "                except KeyError:\n",
    "                    print(\"Using default k = (0.05, 0.06).\")\n",
    "                    k = (0.05,0.06)\n",
    "                finally:\n",
    "                    rate = self.fitness_mutation_rate(k, f_max, f_avg, f = gene.fitness)\n",
    "\n",
    "            elif type == 'rank':\n",
    "                r = index\n",
    "                n = self.pop_size\n",
    "                try:\n",
    "                    p_max = params['p_max']\n",
    "                except KeyError:\n",
    "                    print(\"Using default p_max = 0.08.\")\n",
    "                    p_max = 0.08\n",
    "                finally:\n",
    "                    rate = self.rank_mutation_rate(p_max, r, n)\n",
    "\n",
    "            elif type == 'random':\n",
    "                try:\n",
    "                    rate = params['rate']\n",
    "                except KeyError:\n",
    "                    print(\"Using default rate = 0.01.\")\n",
    "                    rate = 0.01\n",
    "\n",
    "            else:\n",
    "                raise ValueError(\"No such mutation type.\")\n",
    "            \n",
    "            new_gene = gene.mutate(self.param_grid, rate)\n",
    "            new_pop += [new_gene]\n",
    "        \n",
    "        if inplace:\n",
    "            self.population = new_pop\n",
    "            new_pop = self\n",
    "        else:\n",
    "            new_pop = Population(self.X_train, self.X_test, self.y_train, self.y_train, self.estimator, self.param_grid, self.pop_size, self.metric, new_pop)\n",
    "\n",
    "        return new_pop\n",
    "    \n",
    "    def best_solution(self):\n",
    "        return self.population[0]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [],
   "source": [
    "class GenomeGrid():\n",
    "    def __init__(self, estimator, param_grid, max_evol, pop_size, mutation_type, metric = 'mse', el_prop = 0.2, nco_rate = 0.5, **params):\n",
    "        \"\"\"'\n",
    "        Parameters/Attributes\n",
    "        -------------------\n",
    "        estimator: Model\n",
    "        param_grid: Parameters\n",
    "        max_evol: Max evolution\n",
    "        pop_size: Population size\n",
    "        mutation_type: 'fitneses', 'random', 'rank'\n",
    "\n",
    "        optional:\n",
    "        scoring: Evaluation criteria\n",
    "        elitism: Prop of population for elitism, default = 0.2\n",
    "        cross_over_rate/nco_rate: Default = 0.5\n",
    "        type: 'fitness', 'random', 'rank'\n",
    "            Default\n",
    "            - 'fitness' : k = (k1,k2) = (0.05,0.06)\n",
    "            - 'random': rate = 0.5\n",
    "            -'rank': p_max = 0.08\n",
    "        \n",
    "        \"\"\"\n",
    "        self._estimator = estimator\n",
    "        self._param_grid = param_grid\n",
    "        self._max_evol = max_evol\n",
    "        self._pop_size = pop_size\n",
    "        self._mutation_type = mutation_type\n",
    "        self._metric = metric\n",
    "        self._el_prop = el_prop\n",
    "        self._nco_rate = nco_rate\n",
    "        if mutation_type == 'fitness':\n",
    "            self._mutation_para = params.get('k', (0.05, 0.06))\n",
    "        elif mutation_type == 'random':\n",
    "            self._mutation_para = params.get('rate', 0.5)\n",
    "        elif mutation_type == 'rank':\n",
    "            self._mutation_para = params.get('p_max', 0.08)\n",
    "        else:\n",
    "            raise Exception('No such mutation type')\n",
    "\n",
    "    @property\n",
    "    def estimator(self):\n",
    "        return self._estimator\n",
    "\n",
    "    @property\n",
    "    def param_grid(self):\n",
    "        return self._param_grid\n",
    "    \n",
    "    @property\n",
    "    def max_evol(self):\n",
    "        return self._max_evol\n",
    "    \n",
    "    @property\n",
    "    def pop_size(self):\n",
    "        return self._pop_size\n",
    "    \n",
    "    @property\n",
    "    def mutation_type(self):\n",
    "        return self._mutation_type\n",
    "    \n",
    "    @property\n",
    "    def metric(self):\n",
    "        return self._metric\n",
    "    \n",
    "    # @property\n",
    "    # def k(self):\n",
    "    #     if self._k:\n",
    "    #         return self._k\n",
    "    #     else:\n",
    "    #         raise ValueError(\"k tuple not initialised\")\n",
    "        \n",
    "    # @property\n",
    "    # def rate(self):\n",
    "    #     if self._rate:\n",
    "    #         return self._rate\n",
    "    #     else:\n",
    "    #         raise ValueError(\"rate not intialised\")\n",
    "        \n",
    "    # @property\n",
    "    # def p_max(self):\n",
    "    #     if self._p_max:\n",
    "    #         return self._p_max\n",
    "    #     else:\n",
    "    #         raise ValueError(\"p_max not initialised\")\n",
    "\n",
    "    @property\n",
    "    def mutation_para(self):\n",
    "        return self._mutation_para    \n",
    "\n",
    "    @property\n",
    "    def nco_rate(self):\n",
    "        return self._nco_rate\n",
    "\n",
    "    @property\n",
    "    def el_prop(self):\n",
    "        return self._el_prop\n",
    "    \n",
    "    \n",
    "    def train_model(self, X_train, X_test, y_train, y_test, verbose = True):\n",
    "        pop = Population(X_train, X_test, y_train, y_test, self.estimator, self.param_grid, self._pop_size, self.metric)\n",
    "        for i in range(self.max_evol):\n",
    "            \n",
    "            ## Elitism: Choose top t\n",
    "            top_t = pop.elitism(self.el_prop)\n",
    "            parent_pairs = self.pop_size - len(top_t)\n",
    "\n",
    "            ## Selection: Make a list of tuple pairs for crossover from subset\n",
    "            parent_pairs = pop.selection(parent_pairs)\n",
    "\n",
    "            ## Cross Over subset\n",
    "            children = []\n",
    "            for pair in parent_pairs:\n",
    "                child = pop.n_cross_over(pair, 1, self.nco_rate, verbose)\n",
    "                print(f'this is child {pair}: {child}')\n",
    "                children += child\n",
    "            \n",
    "            pop = top_t + children\n",
    "            pop = Population(X_train, X_test, y_train, y_test, self.estimator, self.param_grid, self._pop_size, self.metric, pop)\n",
    "            \n",
    "            ## Mutate everything, including those carried over by elitism\n",
    "            pop.mutation(self.mutation_type, self.mutation_para)          \n",
    "\n",
    "            if verbose:\n",
    "                print(f'--------------Generation {i}------------\\n{pop}')\n",
    "        \n",
    "        best = pop.best_solution()\n",
    "\n",
    "        return best\n",
    "    \n",
    "        \n",
    "\n",
    "            \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "        \n",
    "\n",
    "        \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Test\n",
    "\n",
    "param_grid = {\n",
    "    'a': {'range': [1,10], 'dtype' : 'int'},\n",
    "    'b': {'range': [1,10], 'dtype' : 'int'},\n",
    "    'c': {'range': [1,10], 'dtype' : 'int'},\n",
    "    'd': {'range':['A','B','C'], 'dtype' : 'category'}\n",
    "}\n",
    "\n",
    "## Wrong or edge cases\n",
    "edge_cases = {  \n",
    "    'd': {'range': [1,2], 'dtype': 'int'},      ## Convert to below\n",
    "    'e': {'range': [1], 'dtype' : 'int'},\n",
    "    'd': {'range': [1,2], 'dtype': 'float'}\n",
    "}\n",
    "\n",
    "wrong_cases = {\n",
    "    'a': {'range': [], 'dtype':'int'},\n",
    "    'b': {'range': [1,1], 'dtype': 'int'},\n",
    "    'c': {'range': [1,2,3], 'dtype': 'category'}\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [],
   "source": [
    "param_grid = {\n",
    "            'kernel': {'range' :['linear','poly'], 'dtype': 'category'},\n",
    "            'C':{'range' : [0, 1], 'dtype': 'float'},\n",
    "            'gamma': {'range': [0,1], 'dtype': 'float'}\n",
    "            }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn import svm\n",
    "from sklearn.datasets import load_iris\n",
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "df = load_iris()\n",
    "X_train, X_test, y_train, y_test = train_test_split(df.data, df.target, test_size = 0.2, random_state = 123)\n",
    "\n",
    "\n",
    "model = svm.SVC()\n",
    "\n",
    "pop = GenomeGrid(model, param_grid, 100, 10, mutation_type = 'rank', p_max = 0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pop.train_model(X_train, X_test, y_train, y_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_pop = pop.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_pop.mutation('rank', p_max = 0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is [(6, 0), (5, 1)]\n",
      "p value = 0.76913 is more than 0.000000. No mutation.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(Genome = Params : {'a': 2, 'b': 7, 'c': 3, 'd': 'C'}, Fitness : 79,\n",
       " Genome = Params : {'a': 9, 'b': 4, 'c': 9, 'd': 'A'}, Fitness : 87)"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x = pop.selection(2)\n",
    "print(f'This is {x}')\n",
    "pop.n_cross_over(x[1],n = 2, rate = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "original: Genome = Params : {'a': 6, 'b': 5, 'c': 1, 'd': 'A'}, Fitness : 77\n",
      "After mutation: 'a': '1'\n",
      "After mutation: 'b': '2'\n",
      "After mutation: 'c': '3'\n",
      "After mutation: 'd': 'C'\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Genome = Params : {'a': 1, 'b': 2, 'c': 3, 'd': 'C'}, Fitness : 73"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x = pop[0]\n",
    "x.mutate(param_grid, 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 639,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[2, 8, 9]"
      ]
     },
     "execution_count": 639,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tim_sort(list(np.random.choice([1,2,3,4,5,6,7,8,9,10], size = 3, replace = False)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
